### PR TITLE
feat: [docs for ruby v3 ufc + bandits] (FF-2859)

### PR DESCRIPTION
**Pull request by leoromanovsky**
ruby v3 has been published with the UFC changes and bandits. I have modeled these docs on the python layout.

### CODE REVIEW

File: docs/sdks/server-sdks/ruby.md [hunk 1]
```diff
@@ -19,33 +21,75 @@ gem install eppo-server-sdk
...
 ```
-gem 'eppo-server-sdk', '~> 0.3.0'
+gem 'eppo-server-sdk', '~> 3.0.0'
 ```
 
-## 2. Initialize the SDK
+### B. Initialize the SDK
 
-Initialize the SDK with a SDK key, which can be generated in the Eppo interface. Initialization the SDK when your application starts up to generate a singleton client instance, once per application lifecycle:
+To initialize the SDK, you will need an SDK key. You can generate one [in the flag interface](https://eppo.cloud/feature-flags/keys).
```
Comment (aarsilv) [inline docs/sdks/server-sdks/ruby.md:29]: minor, but would it be better to call it the "flags" interface (plural), or "configuration" as that's what they click to expose the SDK keys tab. OR we could just link to that!

image

File: docs/sdks/server-sdks/ruby.md [hunk 2]
```diff
@@ -19,33 +21,75 @@ gem install eppo-server-sdk
...
+```ruby
+require 'eppo_client'
+
+client = EppoClient::Client.instance
+variation = client.get_string_assignment(
+  '<FLAG-KEY>',
+  '<SUBJECT-KEY>',
+  {
+    # Optional map of subject metadata for targeting.
+  },
```
Comment (aarsilv) [inline docs/sdks/server-sdks/ruby.md:57]: I find `# Optional` confusing as later on you state we need to pass _something_, such as an empty dictionary. Maybe we should rephrase it to read like "Mapping of any subject metadata for targeting"

Note that Some of our other languages have syntactic sugar that let downstream developers skip it entirely, making the _parameter_ optional (as opposed to just its contents). Something we may want to do in rust/ruby land.

Lastly, in Ruby world should we say "hash" instead of "map" 🤔

File: docs/sdks/server-sdks/ruby.md [hunk 3]
```diff
@@ -19,33 +21,75 @@ gem install eppo-server-sdk
...
+  '<DEFAULT-VALUE>'
+)
+```
+
+The `get_string_assignment` function takes four required inputs to assign a variation:
+
+* `<FLAG-KEY>` is the key that you chose when creating a flag; you can find it on the [flag page](https://eppo.cloud/feature-flags). For the rest of this presentation, we'll use `"test-checkout"`. To follow along, we recommend that you create a test flag in your account, and split users between `"fast_checkout"` and `"standard_checkout"`.
+* `<SUBJECT-KEY>` is the value that identifies each entity in your experiment, typically `user_id`.
+* `<SUBJECT-ATTRIBUTES>` is a dictionary of metadata about the subject used for targeting. If you create targeting rules based on attributes, those attributes must be passed in on every assignment call. If no attributes are needed, pass in an empty dictionary.
+* `<DEFAULT-VALUE>` is the value that will be returned if no allocation matches the subject, if the flag is not enabled, if `get_string_assignment` is invoked before the SDK has finished initializing, or if the SDK was not able to retrieve the flag configuration.
```
Comment (aarsilv) [inline docs/sdks/server-sdks/ruby.md:67]: My impression from the slack thread wa that "no allocation matches the subject" is technically an error case given we have default allocations? So perhaps we should leave that out and roll it into a general "or the SDK encountered an error determining the assignment"

File: docs/sdks/server-sdks/ruby.md [hunk 4]
```diff
@@ -65,7 +109,17 @@ config = EppoClient::Config.new(
   '<YOUR_API_KEY>',
   assignment_logger: CustomAssignmentLogger.new
 )
-client = EppoClient::init(config)
+EppoClient::init(config)
+
+# Allow the configuration to be loaded from the CDN.
+sleep(1)
```
Comment (aarsilv) [inline docs/sdks/server-sdks/ruby.md:115]: 🤨 no awaits or callbacks in ruby? or can we make a synchronous version?

Don't love seeing `sleep()` in our example 😞

File: docs/sdks/server-sdks/ruby.md [hunk 5]
```diff
@@ -84,71 +138,152 @@ The SDK will invoke the `log_assignment` function with an `assignment` object th
...
-client = EppoClient::Client.instance
-variation = client.get_string_assignment(
-  '<SUBJECT-KEY>',
-  '<FLAG-KEY>',
-  {
-    # Optional map of subject metadata for targeting.
-  }
-)
-```
+At initialization, the SDK polls Eppo's API to retrieve the most recent experiment configuration. The SDK stores that configuration in memory. This is why assignments are effectively instant, as you can see yourself by profiling the code above.
```
Comment (aarsilv) [inline docs/sdks/server-sdks/ruby.md:147]: technically, it polls our CDN, hydrated from our API. I'ts probably worth calling out that distinction.

File: docs/sdks/server-sdks/ruby.md [hunk 6]
```diff
@@ -84,71 +138,152 @@ The SDK will invoke the `log_assignment` function with an `assignment` object th
...
-```
+### B. Automatically Updating the SDK Configuration
+
+After initialization, the SDK continues polling Eppo's API at 30-second intervals. This retrieves the most recent flag and experiment configurations such as variation values, targeting rules, and traffic allocation. This happens independently of assignment calls.
 
-### Handling `nil`
+:::note
 
-We recommend always handling the `nil` case in your code. Here are some examples of when the SDK returns `nil`:
+Changes made to experiments on Eppo's web interface are almost instantly propagated through our Content-delivery network (CDN) Fastly. Because of the refresh rate, it may take up to 30 seconds (± 5 seconds fuzzing) for those to be reflected by the SDK assignments.
```
Comment (aarsilv) [inline docs/sdks/server-sdks/ruby.md:163]: consider the term "jitter" instead of "fuzzing"

File: docs/sdks/server-sdks/ruby.md [hunk 7]
```diff
@@ -84,71 +138,152 @@ The SDK will invoke the `log_assignment` function with an `assignment` object th
...
+### A. Add a bandit action logger to the assignment logger
+
+In order for the bandit to learn an optimized policy, we need to capture and log the bandit actions.
+This requires adding a bandit action logging callback to the AssignmentLogger class
+```ruby
+class MyLogger < EppoClient::AssignmentLogger
+    def log_assignment(assignment):
+        ...
+
+    def log_bandit_action(self, bandit_action):
```
Comment (aarsilv) [inline docs/sdks/server-sdks/ruby.md:189]: why the `self` argument?

In other SDKs and their documentation, we call the parameters `assignmentEvent` and `banditEvent`, as they contain more than just the assignment and action respectively.

File: docs/sdks/server-sdks/ruby.md [hunk 8]
```diff
@@ -84,71 +138,152 @@ The SDK will invoke the `log_assignment` function with an `assignment` object th
...
+| `flagKey` (String)                                | The key of the feature flag corresponding to the bandit                                                                                           | "bandit-test-allocation-4"          |
+| `banditKey` (String)                                       | The key (unique identifier) of the bandit                                                                       | "ad-bandit-1"                       |
+| `subject` (String)                                   | An identifier of the subject or user assigned to the experiment variation                                       | "ed6f85019080"                      |
+| `action` (String)                                    | The action assigned by the bandit                                                                               | "promo-20%-off"                     |
+| `subjectNumericAttributes` (Hash{String => Float})     | Metadata about numeric attributes of the subject. Hash of the name of attributes their numeric values            | `{"age": 30}`    |
```
Comment (aarsilv) [inline docs/sdks/server-sdks/ruby.md:202]: I feel like we're missing a word here. I think we want something like "hash of the name of the attributes **_to_** their numeric values" or "hash of the name of the attributes **_and_** their numeric values"

File: docs/sdks/server-sdks/ruby.md [hunk 9]
```diff
@@ -84,71 +138,152 @@ The SDK will invoke the `log_assignment` function with an `assignment` object th
...
+| `actionNumericAttributes` (Hash{String => Float})      | Metadata about numeric attributes of the assigned action. Hash of the name of attributes their numeric values    | `{"discount": 0.1}`   |
+| `actionCategoricalAttributes` (Hash{String => String})  | Metadata about non-numeric attributes of the assigned action. Hash of the name of attributes their string values | `{"promoTextColor": "white"}` |
+| `actionProbability` (Float)                         | The weight between 0 and 1 the bandit valued the assigned action                                                | 0.25                                |
+| `modelVersion` (String)                              | Unique identifier for the version (iteration) of the bandit parameters used to determine the action probability | "v123"                       |
+
+### B. Querying the bandit for an action
+
+To query the bandit for an action, you can use the `get_bandit_action` function. This function takes the following parameters:
+- `flag_key` (str): The key of the feature flag corresponding to the bandit
```
Comment (aarsilv) [inline docs/sdks/server-sdks/ruby.md:212]: Rubify to `(String)`?

File: docs/sdks/server-sdks/ruby.md [hunk 10]
```diff
@@ -84,71 +138,152 @@ The SDK will invoke the `log_assignment` function with an `assignment` object th
...
+        # implement me
+```
+
+We automatically log the following data:
+
+| Field                                                | Description                                                                                                     | Example                             |
+|------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------|-------------------------------------|
+| `timestamp` (Date)                                   | The time when the action is taken in UTC  | 2024-03-22T14:26:55.000Z            |
+| `flagKey` (String)                                | The key of the feature flag corresponding to the bandit                                                                                           | "bandit-test-allocation-4"          |
+| `banditKey` (String)                                       | The key (unique identifier) of the bandit                                                                       | "ad-bandit-1"                       |
```
Comment (aarsilv) [inline docs/sdks/server-sdks/ruby.md:199]: does our Ruby event use cammel case property names? Should we make it snake for Ruby land? If we don't have a translation layer, given the difference in language conventions, I wonder if event case is some setting we pass through to rust.

File: docs/sdks/server-sdks/ruby.md [hunk 11]
```diff
@@ -84,71 +138,152 @@ The SDK will invoke the `log_assignment` function with an `assignment` object th
...
+| `actionCategoricalAttributes` (Hash{String => String})  | Metadata about non-numeric attributes of the assigned action. Hash of the name of attributes their string values | `{"promoTextColor": "white"}` |
+| `actionProbability` (Float)                         | The weight between 0 and 1 the bandit valued the assigned action                                                | 0.25                                |
+| `modelVersion` (String)                              | Unique identifier for the version (iteration) of the bandit parameters used to determine the action probability | "v123"                       |
+
+### B. Querying the bandit for an action
+
+To query the bandit for an action, you can use the `get_bandit_action` function. This function takes the following parameters:
+- `flag_key` (str): The key of the feature flag corresponding to the bandit
+- `subject_key` (str): The key of the subject or user assigned to the experiment variation
+- `subject_attributes` (Attributes): The context of the subject
```
Comment (aarsilv) [inline docs/sdks/server-sdks/ruby.md:214]: In other SDKs we allow either freeform attributes ([key] => bool | number | string) or contextual attributes (numeric_attributes => [key] => number, categorical_attributes => [key] => bool | number | string).

Assuming we do here (and if not, we should adjust) we should call this out in the documentation. It may also help to right after this explain the difference, as we do in our JavaScript documentation ([link](https://docs.geteppo.com/sdks/server-sdks/node/#querying-the-bandit-for-an-action)).

File: docs/sdks/server-sdks/ruby.md [hunk 12]
```diff
@@ -84,71 +138,152 @@ The SDK will invoke the `log_assignment` function with an `assignment` object th
...
+| `actionProbability` (Float)                         | The weight between 0 and 1 the bandit valued the assigned action                                                | 0.25                                |
+| `modelVersion` (String)                              | Unique identifier for the version (iteration) of the bandit parameters used to determine the action probability | "v123"                       |
+
+### B. Querying the bandit for an action
+
+To query the bandit for an action, you can use the `get_bandit_action` function. This function takes the following parameters:
+- `flag_key` (str): The key of the feature flag corresponding to the bandit
+- `subject_key` (str): The key of the subject or user assigned to the experiment variation
+- `subject_attributes` (Attributes): The context of the subject 
+- `actions` (Hash{String => Attributes}): A hash that maps available actions to their attributes
```
Comment (aarsilv) [inline docs/sdks/server-sdks/ruby.md:215]: Same for actions, except we also allow just a set of Strings too if we don't have any action context.

File: docs/sdks/server-sdks/ruby.md [hunk 13]
```diff
@@ -84,71 +138,152 @@ The SDK will invoke the `log_assignment` function with an `assignment` object th
...
+Note that action contexts can contain two kinds of information:
+- Action-specific context: e.g., the image aspect ratio of the image corresponding to this action
+- User-action interaction context: e.g., there could be a "brand-affinity" model that computes brand affinities of users to brands, and scores of this model can be added to the action context to provide additional context for the bandit.
+
+#### Result
+
+The `bandit_result` is an instance of `BanditResult`, which has two fields:
+
+- `variation` (String): The variation that was assigned to the subject
+- `action` (Optional[String]): The action that was assigned to the subject
```
Comment (aarsilv) [inline docs/sdks/server-sdks/ruby.md:276]: is `Optional[String]` Ruby-speek?

File: docs/sdks/server-sdks/ruby.md [hunk 14]
```diff
@@ -84,71 +138,152 @@ The SDK will invoke the `log_assignment` function with an `assignment` object th
...
+
+The variation returns the feature flag variation; this can be the bandit itself, or the "status quo" variation if the user is not assigned to the bandit.
+If we are unable to generate a variation, for example when the flag is turned off, then the `default` variation is returned. 
+In both of those cases, the `action` is `nil`, and you should use the status-quo algorithm to select an action.
```
Comment (aarsilv) [inline docs/sdks/server-sdks/ruby.md:280]: this may not make sense because one of the above cases is the bandit itself, in which case `action` won't be `nil`.

Perhaps reword to "In both of those cases, when `action` is `nil`, you should use the status-quo algorithm to select an action.

File: docs/sdks/server-sdks/ruby.md [hunk 15]
```diff
@@ -84,71 +138,152 @@ The SDK will invoke the `log_assignment` function with an `assignment` object th
...
+
+#### Result
+
+The `bandit_result` is an instance of `BanditResult`, which has two fields:
+
+- `variation` (String): The variation that was assigned to the subject
+- `action` (Optional[String]): The action that was assigned to the subject
+
+The variation returns the feature flag variation; this can be the bandit itself, or the "status quo" variation if the user is not assigned to the bandit.
+If we are unable to generate a variation, for example when the flag is turned off, then the `default` variation is returned.
```
Comment (aarsilv) [inline docs/sdks/server-sdks/ruby.md:279]: There is another case, where the bandit encounters an error selecting an action (or there are no actions provided) in which case it will return the bandit variation, but a `nil` action. I wonder if that's worth explicitly calling out--I think it is.

**Review by aarsilv — APPROVED** [review]
Nice work! 🎉  Approving with a BOATLOAD 🛥️ of comments you are welcome to act on or ignore as you see fit!

File: docs/sdks/server-sdks/ruby.md (hunk 4 above)
Comment (felipecsl) [reply docs/sdks/server-sdks/ruby.md:115]: agreed, ruby is single threaded by default though and doesnt have async primitives like async/await. we could have a callback instead, but that would require changing the interface

File: docs/sdks/server-sdks/ruby.md (hunk 7 above)
Comment (felipecsl) [reply docs/sdks/server-sdks/ruby.md:189]: maybe a python-ism?

Comment (leoromanovsky) [reply docs/sdks/server-sdks/ruby.md:189]: Thanks for noticing - sorry I copied this from python and missed this.
